compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -41,306 +43,297 @@ import org.jetbrains.org.objectweb.asm.commons.Method
+        private val jvmSignatureClashDetector = JvmSignatureClashDetector(irClass, type, context)
+        private val smap = context.getSourceMapper(irClass)
 
-        val fieldType = typeMapper.mapType(field)
-        val fieldSignature =
-            if (field.origin == IrDeclarationOrigin.PROPERTY_DELEGATE) null
-            else context.methodSignatureMapper.mapFieldSignature(field)
-        val fieldName = field.name.asString()
-        val fv = visitor.newField(
-            field.OtherOrigin, field.flags, fieldName, fieldType.descriptor,
-            fieldSignature, (field.initializer?.expression as? IrConst<*>)?.value
-        )
+        protected val visitor = state.factory.newVisitor(classOrigin, type, irClass.fileParent.loadSourceFilesInfo(context)).apply {
+            val signature = irClass.getSignature(type, typeMapper)
+            // Ensure that the backend only produces class names that would be valid in the frontend for JVM.
+            if (context.state.classBuilderMode.generateBodies && signature.hasInvalidName()) {
+                throw IllegalStateException("Generating class with invalid name '${type.className}': ${irClass.dump()}")
+            }
+            defineClass(
+                classOrigin.element,
+                state.classFileVersion,
+                irClass.flags,
+                signature.name,
+                signature.javaGenericSignature,
+                signature.superclassName,
+                signature.interfaces.toTypedArray()
+            )
+        }
 
-        jvmSignatureClashDetector.trackField(field, RawSignature(fieldName, fieldType.descriptor, MemberKind.FIELD))
+        fun generate() {
+            // Generating a method node may cause the addition of a field with an initializer if an inline function
+            // uses `assert` and the JVM assertion mode is enabled. To avoid concurrent modification errors,
+            // there is a very specific member generation order (the JVM doesn't care).
+            // 1. Any method other than `<clinit>` can add a field and a `<clinit>` statement:
+            for (method in irClass.declarations.filterIsInstance<IrFunction>()) {
+                if (method.name.asString() != "<clinit>") {
+                    generateMethod(method)
+                }
+            }
+            // 2. `<clinit>` can add a field (the statement is generated inline via the `return init` hack;
+            //    see `generateAssertFieldIfNeeded`):
+            val clinit = irClass.functions.singleOrNull { it.name.asString() == "<clinit>" }
+            if (clinit != null) {
+                generatingClInit = true
+                generateMethod(clinit)
+            }
+            // 3. Now we have all the fields:
+            for (declaration in irClass.fields) {
+                generateField(declaration)
+            }
+            // 4. Generate inner classes at the end, to ensure that when the companion's metadata is serialized
```

Comment:
The previously used term "nested" is more correct here since it includes classes both with and without `this$0` (see https://kotlinlang.org/docs/reference/nested-classes.html)